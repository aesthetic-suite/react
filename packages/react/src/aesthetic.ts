import { Aesthetic } from '@aesthetic/core';
import { createClientEngine } from '@aesthetic/style';

const aesthetic = new Aesthetic();

aesthetic.configureEngine(createClientEngine());

export default aesthetic;
