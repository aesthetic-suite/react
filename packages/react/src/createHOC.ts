/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { WrapperComponent } from './types';

export default function createHOC<T extends React.FunctionComponent<any>>(
  name: string,
  WrappedComponent: React.ComponentType<any>,
  BaseComponent: T,
): T & WrapperComponent {
  const Component = BaseComponent as T & WrapperComponent;

  hoistNonReactStatics(Component, WrappedComponent);

  Component.displayName = `${name}(${WrappedComponent.displayName || WrappedComponent.name})`;
  Component.WrappedComponent = WrappedComponent;

  return Component;
}
