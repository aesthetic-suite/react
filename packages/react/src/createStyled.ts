import React from 'react';
import { createComponentStyles, LocalBlock, Utilities } from '@aesthetic/core';
import useStyles from './useStyles';

export default function createStyled<T extends keyof JSX.IntrinsicElements>(
  type: T,
  factory: (utilities: Utilities<LocalBlock>) => LocalBlock,
): React.ForwardRefExoticComponent<JSX.IntrinsicElements[T]> {
  if (__DEV__) {
    if (typeof factory !== 'function') {
      throw new TypeError(
        `Styled components require a style sheet factory function, found ${typeof factory}.`,
      );
    }
  }

  const styleSheet = createComponentStyles((utils) => ({
    element: factory(utils),
  }));

  const Component = React.forwardRef<unknown, JSX.IntrinsicElements[T]>((props, ref) => {
    const cx = useStyles(styleSheet);
    let className = cx('element');

    if (props.className) {
      className += ` ${props.className}`;
    }

    return React.createElement(type, {
      ...props,
      className,
      ref,
    });
  });

  Component.displayName = `styled(${type})`;

  return Component;
}
