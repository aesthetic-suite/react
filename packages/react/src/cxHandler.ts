import { ClassNameSheet, ClassNameSheetVariants, generateClassName } from '@aesthetic/core';
import { isObject, objectLoop } from '@aesthetic/utils';

const variantCache: Record<string, string[]> = {};

export default function cxHandler(keys: unknown[], classNames: ClassNameSheet<string>) {
  const names = keys as string[];
  let variantNames: string[] = [];

  if (isObject(names[0])) {
    const variants = (names.shift() as unknown) as ClassNameSheetVariants;
    const cacheKey = JSON.stringify(variants);
    const cache = variantCache[cacheKey];

    if (cache) {
      variantNames = cache;
    } else {
      objectLoop(variants, (value, type) => {
        variantNames.push(`${type}_${value}`);
      });

      variantCache[cacheKey] = variantNames;
    }
  }

  return generateClassName(names, variantNames, classNames);
}
