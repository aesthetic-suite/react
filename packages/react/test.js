module.exports = require('@aesthetic/core/test');
