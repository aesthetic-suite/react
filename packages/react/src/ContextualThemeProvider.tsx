import React, { useState, useEffect } from 'react';
import aesthetic from './aesthetic';
import ThemeContext from './ThemeContext';
import useDirection from './useDirection';
import { ContextualThemeProviderProps } from './types';

/**
 * Provides a theme for a distinct portion of the React tree.
 */
export default function ContextualThemeProvider({
  children,
  name,
  wrapper,
}: ContextualThemeProviderProps) {
  const [className, setClassName] = useState('');
  const direction = useDirection();
  const theme = aesthetic.getTheme(name);

  // Render styles when theme/direction change
  useEffect(() => {
    setClassName(aesthetic.renderThemeStyles(theme, { direction }));
  }, [theme, direction]);

  return (
    <ThemeContext.Provider value={theme}>
      {React.cloneElement(wrapper, { className, 'data-theme': name }, children)}
    </ThemeContext.Provider>
  );
}
