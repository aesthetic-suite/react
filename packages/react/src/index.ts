/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { aesthetic } from './aesthetic';
import { createStyled } from './styles';

export * from './direction';
export * from './styles';
export * from './theme';
export * from './types';
export * from '@aesthetic/core';
export * from '@aesthetic/core-react';

export const {
	changeDirection,
	changeTheme,
	configure,
	createStyleSheet,
	createScopedStyleSheet,
	createThemeSheet,
	getActiveDirection,
	getActiveTheme,
	getEngine,
	getTheme,
	registerDefaultTheme,
	registerTheme,
	renderStyleSheet,
	renderFontFace,
	renderImport,
	renderKeyframes,
	renderThemeSheet,
	subscribe,
	unsubscribe,
} = aesthetic;

// Export aliases
export const style = createScopedStyleSheet;
export const styles = createStyleSheet;
export const styled = createStyled;
export const themeStyle = createThemeSheet;

// Do not use! This is for bundler pre-compilation usage!
// @private
export const internalAestheticRuntime = aesthetic;
