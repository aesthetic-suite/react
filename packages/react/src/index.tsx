/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import aesthetic from './aesthetic';
import ThemeContext from './ThemeContext';
import ThemeProvider from './ThemeProvider';
import ContextualThemeProvider from './ContextualThemeProvider';
import createStyled from './createStyled';
import useStyles from './useStyles';
import useTheme from './useTheme';
import withStyles from './withStyles';
import withTheme from './withTheme';

export * from '@aesthetic/core';
export * from './direction';
export * from './types';

export {
  ThemeContext,
  ThemeProvider,
  ContextualThemeProvider,
  createStyled,
  useStyles,
  useTheme,
  withStyles,
  withTheme,
};

export const {
  changeDirection,
  changeTheme,
  configure,
  createComponentStyles,
  createThemeStyles,
  generateClassName,
  getActiveDirection,
  getActiveTheme,
  getEngine,
  getTheme,
  registerDefaultTheme,
  registerTheme,
  renderComponentStyles,
  renderFontFace,
  renderImport,
  renderKeyframes,
  renderThemeStyles,
  subscribe,
  unsubscribe,
} = aesthetic;
