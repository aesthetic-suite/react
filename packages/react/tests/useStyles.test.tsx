/* eslint-disable react/jsx-no-literals */

import React from 'react';
import { render } from 'rut-dom';
import { getRenderedStyles } from '@aesthetic/style/test';
import { useStyles } from '../src';
import { ButtonProps, createStyleSheet, Wrapper } from './__mocks__/Button';
import { dawnTheme, setupAestheticReact, teardownAestheticReact, twilightTheme } from './helpers';

describe('useStyles()', () => {
  function Button({ children, block, disabled, large, small }: ButtonProps) {
    const cx = useStyles(createStyleSheet());

    return (
      <button
        type="button"
        className={cx(
          {
            // eslint-disable-next-line no-nested-ternary
            size: large ? 'lg' : small ? 'sm' : 'df',
          },
          'button',
          block && 'button_block',
          disabled && 'button_disabled',
        )}
      >
        {children}
      </button>
    );
  }

  function CustomSheetButton({ children, sheet }: ButtonProps) {
    const cx = useStyles(sheet!);

    return (
      <button type="button" className={cx()}>
        {children}
      </button>
    );
  }

  function StylelessButton({ children, disabled }: ButtonProps) {
    const cx = useStyles(createStyleSheet());

    return (
      <button type="button" className={cx(disabled && 'button_disabled')}>
        {children}
      </button>
    );
  }

  beforeEach(() => {
    setupAestheticReact();
  });

  afterEach(() => {
    teardownAestheticReact();
  });

  it('renders a button and its base styles', () => {
    const { root } = render<ButtonProps>(<Button>Child</Button>, {
      wrapper: <Wrapper />,
    });

    expect(getRenderedStyles('standard')).toMatchSnapshot();

    expect(root.findOne('button')).toHaveProp(
      'className',
      'a b c d e f g h i j k l m n o p q r s t u v w z a1',
    );
  });

  it('changes class name based on props enabled', () => {
    const { root, update } = render<ButtonProps>(<Button>Child</Button>, {
      wrapper: <Wrapper />,
    });

    expect(root.findOne('button')).toHaveProp(
      'className',
      'a b c d e f g h i j k l m n o p q r s t u v w z a1',
    );

    update({ disabled: true });

    expect(root.findOne('button')).toHaveProp(
      'className',
      'a b c d e f g h i j k l m n o p q r s t u v w z a1 l1',
    );

    update({ block: true, large: true });

    expect(root.findOne('button')).toHaveProp(
      'className',
      'a b c d e f g h i j k l m n o p q r s t u v w d1 e1 f1 g1 h1 i1 j1 b1 c1 m1',
    );
  });

  it('returns an empty string if no selectors enabled', () => {
    const { root } = render<ButtonProps>(<StylelessButton>Child</StylelessButton>, {
      wrapper: <Wrapper />,
    });

    expect(root.findOne('button')).toHaveProp('className', '');
  });

  it('only renders once unless theme or direction change', () => {
    const sheet = createStyleSheet();
    const spy = jest.spyOn(sheet, 'render');

    const { update } = render<ButtonProps>(
      <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>,
      {
        wrapper: <Wrapper />,
      },
    );

    update();
    update();
    update();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('re-renders if direction changes', () => {
    const sheet = createStyleSheet();
    const spy = jest.spyOn(sheet, 'render');

    const { rerender } = render<ButtonProps>(
      <Wrapper>
        <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      twilightTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'twilight',
      }),
    );

    rerender(
      <Wrapper direction="rtl">
        <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      twilightTheme,
      expect.objectContaining({
        direction: 'rtl',
        theme: 'twilight',
      }),
    );
  });

  it('re-renders if theme changes', () => {
    const sheet = createStyleSheet();
    const spy = jest.spyOn(sheet, 'render');

    const { rerender } = render<ButtonProps>(
      <Wrapper>
        <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      twilightTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'twilight',
      }),
    );

    rerender(
      <Wrapper theme="dawn">
        <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      dawnTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'dawn',
      }),
    );
  });

  it('re-renders when direction and theme change', () => {
    const sheet = createStyleSheet();
    const spy = jest.spyOn(sheet, 'render');

    const { rerender } = render<ButtonProps>(
      <Wrapper>
        <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      twilightTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'twilight',
      }),
    );

    rerender(
      <Wrapper direction="rtl" theme="dawn">
        <CustomSheetButton sheet={sheet}>Child</CustomSheetButton>
      </Wrapper>,
    );

    // Annoying that this fires multiple times, possible to batch?
    expect(spy).toHaveBeenCalledTimes(3);

    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      twilightTheme,
      expect.objectContaining({
        direction: 'rtl',
        theme: 'twilight',
      }),
    );

    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      dawnTheme,
      expect.objectContaining({
        direction: 'rtl',
        theme: 'dawn',
      }),
    );
  });
});
