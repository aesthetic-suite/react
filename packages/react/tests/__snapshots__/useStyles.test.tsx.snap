// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyles() renders a button and its base styles 1`] = `".a {appearance: none;}.b {background-color: var(--palette-brand-color-40);}.c {border: 0;}.d {cursor: pointer;}.e {display: inline-flex;}.f {font-size: inherit;}.g {margin: 0;}.h {padding: 0;}.i {text-decoration: none;}.j {user-select: auto;}.k {vertical-align: middle;}.l {position: relative;}.m {text-align: center;}.n {color: var(--palette-neutral-color-00);}.o {border-color: var(--palette-neutral-color-40);}.p {border-style: solid;}.q {border-width: var(--border-df-width);}.r {border-radius: var(--border-df-radius);}.s:hover {background-color: var(--palette-brand-color-50);}.t:active {background-color: var(--palette-brand-color-60);}.u::-moz-focus-inner {border: 0;}.v::-moz-focus-inner {padding: 0;}.w::-moz-focus-inner {margin: 0;}.x {min-width: 7.50rem;}.y {padding: var(--spacing-sm) var(--spacing-df);}.z {min-width: 10rem;}.a1 {padding: var(--spacing-df) var(--spacing-md);}.b1 {min-width: 12.50rem;}.c1 {padding: var(--spacing-md) var(--spacing-lg);}.d1 {overflow-wrap: normal;}.e1 {word-wrap: normal;}.f1 {word-break: normal;}.g1 {display: block;}.h1 {width: 100%;}.i1 {white-space: normal;}.j1 {overflow: hidden;}.k1 {opacity: 0.4;}.l1 {opacity: 0.5;}.m1 {opacity: 0.6;}"`;
